CONTRIBUTING.md

```diff
@@ -29,6 +29,50 @@ submitted for inclusion in the work by you, as defined in the
 Apache-2.0 license, shall be dual licensed as in the README, without any
 additional terms or conditions.
 
+## Version numbers
+
+We use `git-testament` to construct our version strings. This includes into
+the binary details of the git commit, tag description, and also an indication
+of modifications to the working tree present when the binary was compiled.
+
+This means that in normal development, you might get information from invoking
+`rustup --version` which look like `rustup-init 1.18.3+15 (a54051502 2019-05-26)`
+or even like `rustup-init 1.18.3+15 (a54051502 2019-05-26) dirty 1 modification`.
+
+The first part is always the binary name as per `clap`'s normal operation. The
+version number is a combination of the most recent tag in the git repo, and the
+number of commits since that tag. The parenthesised information is, naturally,
+the SHA of the most recent commit and the date of that commit. If the indication
+of a dirty tree is present, the number of changes is indicated.  This combines
+adds, deletes, modifies, and unknown entries.
+
+You can request further information of a `rustup` binary with the
+`rustup dump-testament` hidden command.  It produces output of the form:
+
+```shell
+$ rustup dump-testament
+Rustup version renders as: 1.18.3+15 (a54051502 2019-05-26) dirty 1 modification
+Current crate version: 1.18.3
+Built from branch: kinnison/version-strings
+Commit info: 1.18.3+15 (a54051502 2019-05-26)
+Modified: CONTRIBUTING.md
+```
+
+This can be super-handy when you are testing development versions on your PC
```

Comment:
Personally I find the superlative here a little jarring, but whats life without a little fun :). Just mentioning it for you to consider. I find it interesting that you do this, because I literally never install rustup when testing it, except for testing the proxies and self-update code I guess...

Reply:
Dropped the superlative, added a mention of users having non-release builds:

```markdown
This can be handy when you are testing development versions on your PC
and cannot remember exactly which version you had installed, or if you have given
a development copy (or instruction to build such) to a user, and wish to have them
confirm *exactly* what they are using.
```